### PR TITLE
Explicit simd for ad

### PR DESCRIPTION
**Pull request by KristofferC**
Unfortunately no way to add this without copy pasting a bunch of code from ForwardDiff.

Benchmarks (on my laptop): https://gist.github.com/KristofferC/9b62f7c130cc4eef256092c363c8a01c

### CODE REVIEW

File: src/partials.jl [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+@generated function minus_tuple{N}(tup::NTuple{N})
+    return tupexpr(i -> :(-tup[$i]), N)
+end
+
+@generated function mul_tuples{N}(a::NTuple{N}, b::NTuple{N}, afactor, bfactor)
+    return tupexpr(i -> :((afactor * a[$i]) + (bfactor * b[$i])), N)
+end
+
+########
+# SIMD #
```
Comment (KristofferC) [inline src/partials.jl]: This and below is the only thing we actually need.

**Comment by KristofferC** [comment]
SIMD O3 vs SIMD: https://gist.github.com/KristofferC/95f7e2c4ca0dc70353858cce06b27ea4
SIMD O3 vs O3: https://gist.github.com/KristofferC/ff54ad0193b62211805b8df1de2fc877
SIMD vs O3: https://gist.github.com/KristofferC/0f8cd24b15a5c1c163d2c3447df09a85

Basically, O3 beats SIMD (with default optimization). SIMD with O3 beats both O3 and SIMD without O3.

**Comment by KristofferC** [comment]
A conclusion here is that it is not good to measure O3 vs SIMD. We should measure SIMD O3 vs O3.

**Comment by fredrikekre** [comment]
Nice! What parts of the code did you add yourself? Just the SIMD section in partials.jl?
This seems like an improvement for the ForwardDiff package too? Perhaps it could be upstreamed?

**Comment by KristofferC** [comment]
Only the SIMD stuff. We are just lucky that we do not hit the bad cases (length 7 for example). This is why this can't really be upstreamed.

File: src/partials.jl [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+@inline function scale_tuple{N, T <: SIMDTypes}(tup::NTuple{N, T}, x)
+    return totuple(tosimd(tup) * x)
+end
+
+@inline function div_tuple_by_scalar{N, T <: SIMDTypes}(tup::NTuple{N, T}, x)
+    return totuple(tosimd(tup) / x)
+end
+
+@inline function add_tuples{N, T1 <: SIMDTypes, T2 <: SIMDTypes}(a::NTuple{N, T1}, b::NTuple{N, T2})
```
Comment (fredrikekre) [inline src/partials.jl]: Need to limit this and `sub_tuples` below to the same type `T`:

```julia
julia> v1 = Vec{4, Float64}((rand(4)...));

julia> v2 = Vec{4, Float32}((rand(Float32, 4)...));

julia> v1 + v2
4-element Array{Float64,1}:
 1.26787
 1.376  
 1.27875
 1.11322
```

Comment (KristofferC) [inline src/partials.jl]: I don't see what is bad about that?

**Comment by KristofferC** [comment]
Rip code coverage...

File: src/partials.jl (hunk 2 above)
Comment (fredrikekre) [inline src/partials.jl]: It turns into an array?
```julia
julia> @which v1 + v2
+{R,S}(A::AbstractArray{R,N<:Any}, B::AbstractArray{S,N<:Any}) at arraymath.jl:63
```

Comment (KristofferC) [inline src/partials.jl]: Ooh!

Comment (fredrikekre) [inline src/partials.jl]: Could fix that in SIMD.jl

**Comment by KristofferC** [comment]
Can you try:

```
import Tensors.Dual
using BenchmarkTools
x, y, z = Dual(rand(9)...) , Dual(rand(9)...), Dual(rand(9)...);
fma2(x,y,z) = x*y + z
@benchmark muladd($x, $y, $z)
@benchmark fma2($x, $y, $z)
```

on your fancy new ass CPU?

**Comment by fredrikekre** [comment]
On this branch?

**Comment by fredrikekre** [comment]
```julia
julia> @benchmark muladd($x, $y, $z)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     2.406 ns (0.00% GC)
  median time:      2.417 ns (0.00% GC)
  mean time:        2.441 ns (0.00% GC)
  maximum time:     18.330 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000

julia> @benchmark fma2($x, $y, $z)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     7.638 ns (0.00% GC)
  median time:      7.641 ns (0.00% GC)
  mean time:        7.750 ns (0.00% GC)
  maximum time:     25.681 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999
```

**Comment by KristofferC** [comment]
Cool. When u have time can u also benchmark the AD stuff and compare it to master.

**Comment by fredrikekre** [comment]
https://gist.github.com/fredrikekre/4a4887eaccd617b79337583c838a616f

**Comment by KristofferC** [comment]
So this PR is crap?

**Comment by fredrikekre** [comment]
The analytical benchmarks in there should not be affected by this PR, right? Even they show bad results...

> So this PR is crap?

Perhaps simd should be limited to dim = 3 atleast?

**Comment by fredrikekre** [comment]
By the way, we are not benchmarking with symmetric tensors due to https://github.com/KristofferC/Tensors.jl/blob/7728952ee6fdcc834c61205f6ac628cee556e68c/benchmark/benchmark_ad.jl#L35 not sure if intended?

**Comment by fredrikekre** [comment]
Btw, seems to be commits from #19 on this branch...

**Comment by fredrikekre** [comment]
New run:
https://gist.github.com/fredrikekre/8569d885c34556120e666adb4f0ac94a
Idk what was up with last run.

**Comment by fredrikekre** [comment]
This PR does nothing now?

**Comment by KristofferC** [comment]
Must have messed up a rebase...

**Comment by KristofferC** [comment]
Some pretty nasty compilation time changes... :(

**Comment by KristofferC** [comment]
Let's try this

**Comment by KristofferC** [comment]
Not inlining some functions when using dim = 3 and dual numbers helped a bit. Just need to check that benchmarks were not affected.

**Comment by KristofferC** [comment]
Turns out I didn't need to import all of ForwardDiff because we specify stricter version of the functions lol...

**Comment by fredrikekre** [comment]
This PR is a roller coaster :roller_coaster: . I feel like every second commit deleted thousand lines and the next one adds them back xD

**Comment by KristofferC** [comment]
I dunno wtf is going on with these rebases...

**Comment by fredrikekre** [comment]
Which side is this PR?

**Comment by KristofferC** [comment]
Right

**Comment by fredrikekre** [comment]
This is also interesting: with:

```julia
julia> A = rand(Tensor{2, 3}); B = symmetric(A);
```

I get on master:
```julia
julia> @btime dot($A, $B);
  7.120 ns (0 allocations: 0 bytes)
```

and with this PR (which changed to muladd):

```julia
julia> @btime dot($A, $B);
  4.734 ns (0 allocations: 0 bytes)
```

 I am sure it was slower with `muladd` for `dot`, thats why it wasn't enabled...

**Comment by KristofferC** [comment]
SIMD.jl is now used upstream in ForwardDiff so this can be closed.

**Pull request closed without merging**